src/core/Internal.ml

```diff
@@ -0,0 +1,2155 @@
+    Log.debugf info (fun k->k "(@[sat.th.add-clause@ %a@])" Clause.debug c);
+    Vec.push st.clauses_to_add c
+
+  let acts_raise st (l:formula list) proof : 'a =
+    let atoms = List.rev_map (create_atom st) l in
+    (* conflicts can be removed *)
+    let c = Clause.make_removable atoms (Lemma proof) in
+    raise_notrace (Th_conflict c)
+
+  let acts_propagate (st:t) f = function
+    | Solver_intf.Eval l ->
+      let a = mk_atom st f in
+      enqueue_semantic st a l
+    | Solver_intf.Consequence (causes, proof) ->
+      let l = List.rev_map (mk_atom st) causes in
+      if List.for_all (fun a -> a.is_true) l then (
+        let p = mk_atom st f in
+        let c = Clause.make_removable (p :: List.map Atom.neg l) (Lemma proof) in
+        if p.is_true then ()
+        else if p.neg.is_true then (
+          Vec.push st.clauses_to_add c
+        ) else (
+          insert_subterms_order st p.var;
+          let level = List.fold_left (fun acc a -> max acc a.var.v_level) 0 l in
+          enqueue_bool st p ~level (Bcp c)
+        )
+      ) else (
+        invalid_arg "Msat.Internal.slice_propagate"
+      )
+
+  let[@specialise] acts_iter st ~full head f : unit =
+    for i = (if full then 0 else head) to Vec.size st.trail-1 do
+      let e = match Vec.get st.trail i with
+        | Atom a ->
+          Solver_intf.Lit a.lit
+        | Lit {term; assigned = Some v; _} ->
+          Solver_intf.Assign (term, v)
+        | Lit _ -> assert false
+      in
+      f e
+    done
+
+  let eval_atom_ a =
+    if Atom.is_true a then Solver_intf.L_true
+    else if Atom.is_false a then Solver_intf.L_false
+    else Solver_intf.L_undefined
+
+  let[@inline] acts_eval_lit st (f:formula) : Solver_intf.lbool =
+    let a = create_atom st f in
+    eval_atom_ a
+
+  let[@inline] acts_mk_lit st f : unit =
+    ignore (create_atom st f : atom)
+
+  let[@inline] acts_mk_term st t : unit = make_term st t
+
+  let[@inline] current_slice st : _ Solver_intf.acts = {
+    Solver_intf.
+    acts_iter_assumptions=acts_iter st ~full:false st.th_head;
+    acts_eval_lit= acts_eval_lit st;
+    acts_mk_lit=acts_mk_lit st;
+    acts_mk_term=acts_mk_term st;
+    acts_add_clause = acts_add_clause st;
+    acts_propagate = acts_propagate st;
+    acts_raise_conflict=acts_raise st;
+  }
+
+  (* full slice, for [if_sat] final check *)
+  let[@inline] full_slice st : _ Solver_intf.acts = {
+    Solver_intf.
+    acts_iter_assumptions=acts_iter st ~full:true st.th_head;
+    acts_eval_lit= acts_eval_lit st;
+    acts_mk_lit=acts_mk_lit st;
+    acts_mk_term=acts_mk_term st;
+    acts_add_clause = acts_add_clause st;
+    acts_propagate = acts_propagate st;
+    acts_raise_conflict=acts_raise st;
+  }
+
+  (* Assert that the conflict is indeeed a conflict *)
+  let check_is_conflict_ (c:Clause.t) : unit =
+    if not @@ Array.for_all (Atom.is_false) c.atoms then (
+      let msg =
+        Format.asprintf "msat:core/internal: invalid conflict %a" Clause.debug c
```

Comment:
One example of error message that doesn't have the same format as the one above.